Sources/Features/Client/BacktraceOomWatcher.swift

```diff
@@ -0,0 +1,179 @@
+    init(repository: PersistentRepository<BacktraceReport>, crashReporter: CrashReporting, attributes: AttributesProvider, backtraceApi: BacktraceApi) {
+        self.crashReporter = crashReporter
+        self.attributesProvider = attributes
+        self.backtraceApi = backtraceApi
+        self.repository = repository
+        
+        // set default state
+        state = ApplicationInfo(
+            attributes: nil,
+            resource: nil)
+        
+        // set status file url
+        if BacktraceOomWatcher.oomFilePath == nil {
+            // database path will point out sqlite database path
+            // oom status should exist in the same directory where database exists.
+            BacktraceOomWatcher.oomFilePath = self.repository.url.deletingLastPathComponent().absoluteURL
+                .appendingPathComponent("BacktraceOOMState.plist")
+        }
+    }
+    
+    internal static func clean() {
+        if let oomFilePath = BacktraceOomWatcher.oomFilePath {
+            // ignore errors or use do/catch block to handle errors more gracefully
+            try? FileManager.default.removeItem(at: oomFilePath)
+        }
+    }
+    
+    public func start() {
+        sendPendingOomReports()
+        // override previous application state after reading all information
+        saveState()
+    }
+    
+    internal static func getStatusFilePath() -> URL? {
+        // oom status file is available in application cache dir - the same dir
+        // where we store attributes
+        guard let cacheDir = FileManager.default.urls(for: .cachesDirectory, in: .userDomainMask).first else {
+            return nil
+        }
+        return cacheDir.appendingPathComponent("BacktraceOomState.plist")
+    }
+    
+    private func sendPendingOomReports() {
+        
+        // if oom state file doesn't exist it means that we deleted it to
+        // prevent sending false oom crashes
+        if !FileManager.default.fileExists(atPath: BacktraceOomWatcher.oomFilePath!.path) {
+            return
+        }
+        
+        guard let appState = self.loadPreviousState() else {
+            return
+        }
+        
+        // no low memory warning
+        if appState.resource == nil {
+            return
+        }
+        
+        // check if debugger was enabled
+        if appState.debugger == true {
+            // detected debugger in previous session
+            return
+        }
+        // check system update
+        if appState.version != ProcessInfo.processInfo.operatingSystemVersionString {
+            // detected system update
+            return
+        }
+        
+        // check application update
+        if appState.appVersion != Bundle.main.object(forInfoDictionaryKey: kCFBundleVersionKey as String) as! String {
+            // detected app update
+            return
+        }
+        
+        let attributes = try? JSONSerialization.jsonObject(with: appState.attributes!, options: []) as! Attributes
```

Comment:
I suggest moving attributes serialization and report creation into the `do/catch` block and removing force unwrapping

Reply:
I have mixed feelings about it - if report is available but JSONSerializer throws an exception then in worst case we won't send useful data for user and fail because of optional values.

Reply:
ok, fine